README.md
# Hashids2
[![Build Status](https://travis-ci.org/olexale/hashids.svg?branch=master)](https://travis-ci.org/olexale/hashids) [![Coverage Status](https://coveralls.io/repos/github/olexale/hashids/badge.svg?branch=master)](https://coveralls.io/github/olexale/hashids?branch=master) [![pub package](https://img.shields.io/pub/v/hashids2.svg)](https://pub.dartlang.org/packages/hashids2)

A Dart class to generate YouTube-like hashes from one or many numbers.

Ported from javascript [hashids.js](https://github.com/ivanakimov/hashids.js) by [Ivan Akimov](https://github.com/ivanakimov)

## What is it?

Hashids (Hash ID's) creates short, unique, decryptable hashes from unsigned (long) integers.

This algorithm tries to satisfy the following requirements:

1. Hashes must be unique and decryptable.
2. They should be able to contain more than one integer (so you can use them in complex or clustered systems).
3. You should be able to specify minimum hash length.
4. Hashes should not contain basic English curse words (since they are meant to appear in public places - like the URL).

Instead of showing items as `1`, `2`, or `3`, you could show them as `U6dc`, `u87U`, and `HMou`.
You don't have to store these hashes in the database, but can encrypt + decrypt on the fly.

All (long) integers need to be greater than or equal to zero.

## Usage

#### Import the package
Follow instructions from [Pub](https://pub.dev/packages/hashids2#-installing-tab-).

#### Encrypting
```dart
final hashids = HashIds();
final fromNumber = hashids.encode(42);
final fromList = hashids.encode([1,2,3]);
final fromString = hashids.encode('42');
```
#### Decrypting
```dart
final hashids = HashIds();
final number = hashids.decode(fromNumber); // [42]
final list = hashids.decode(fromList); // [1,2,3]
final string = hashids.decode(fromString); // [42]
```

## License
MIT License. See the LICENSE file. You can use Hashids in open source projects and commercial products. Don't break the Internet. Kthxbye.


CHANGELOG.md
# Changelog

## 2.0.0

* Migration to null safety

## 1.0.3+1

* Improve pub points (by @sidrao2006)

## 1.0.3

* Fix the original LICENSE as all hashids must be available under MIT

## 1.0.2

* Docuemntation added
* Dead code removed
* More unit tests

## 1.0.1

* Add basic example
* Use named parameters

## 1.0.0

* Ported to Dart 2


test/sample.dart
import 'package:hashids2/hashids2.dart';
import 'package:test/test.dart';

void sampleTest() {
  group('real', () {
    test('Real Sample Hash', () {
      final hashids = HashIds('this is my salt', 8,
          'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890');
      final id = hashids.encode([1, 2, 3]);
      final numbers = hashids.decode(id);
      expect(id, equals('GlaHquq0'));
      expect(numbers, equals([1, 2, 3]));
    });
  });
}


test/all_test.dart
import 'bad_input.dart';
import 'custom_alphabet.dart';
import 'custom_params.dart';
import 'custom_params_hex.dart';
import 'custom_salt.dart';
import 'default_params.dart';
import 'default_params_hex.dart';
import 'encode_types.dart';

void main() {
  badInputTest();
  customSaltTest();
  customAlphabetTest();
  customParamsTest();
  customParamsHexTest();
  defaultParamsTest();
  defaultParamsHexTest();
  encodeTypesTest();
}


test/bad_input.dart
import 'package:hashids2/hashids2.dart';
import 'package:test/test.dart';

void badInputTest() {
  group('bad input', () {
    test('should throw an error when small alphabet', () {
      try {
        HashIds(alphabet: '1234567890');
      } catch (e) {
        expect(1, 1);
      }
    });

    test('should throw an error when alphabet has spaces', () {
      try {
        HashIds(alphabet: 'a cdefghijklmnopqrstuvwxyz');
      } catch (e) {
        expect(1, 1);
      }
    });
    final hashids = HashIds();
    test('should return an empty string when encoding nothing', () {
      final id = hashids.encode(<int>[]);
      expect(id, equals(''));
    });
    test('should return an empty string when encoding a negative number', () {
      final id = hashids.encode([-1]);
      expect(id, equals(''));
    });

    test(
        'should return an empty string when encoding a string with non-numeric characters',
        () {
      expect(hashids.encode('6B'), '');
      expect(hashids.encode('123a'), '');
    });

    test('should return an empty string when encoding infinity', () {
      final id = hashids.encode([double.infinity]);
      expect(id, equals(''));
    });

    test('should return an empty string when encoding a null', () {
      final id = hashids.encode(null);
      expect(id, equals(''));
    });

    test(
        'should return an empty string when encoding an array with non-numeric input',
        () {
      final id = hashids.encode(['z']);
      expect(id, '');
    });

    test('should return an empty array when decoding nothing', () {
      final numbers = hashids.decode('');
      expect(numbers, equals(<int>[]));
    });

    test('should return an empty string when encoding non-numeric input', () {
      final id = hashids.encode('z');
      expect(id, '');
    });

    test('should return an empty array when decoding invalid id', () {
      final numbers = hashids.decode('f');
      expect(numbers, equals(<int>[]));
    });

    test('should return an empty string when encoding non-hex input', () {
      final id = hashids.encodeHex('z');
      expect(id, equals(''));
    });

    test('should return an empty array when hex-decoding invalid id', () {
      final numbers = hashids.decodeHex('f');
      expect(numbers, equals(''));
    });
  });
}


lib/hashids.dart
import 'dart:core';

class HashIds {
  HashIds(
      [String salt = DEFAULT_SALT,
      int minHashLength = DEFAULT_MIN_HASH_LENGTH,
      String alphabet = DEFAULT_ALPHABET]) {
    this.salt = salt != null ? salt : DEFAULT_SALT;
    this.minHashLength =
        minHashLength > 0 ? minHashLength : DEFAULT_MIN_HASH_LENGTH;

    var uniqueAlphabet = '';
    for (int i = 0; i < alphabet.length; i++) {
      if (!uniqueAlphabet.contains(alphabet[i])) {
        uniqueAlphabet += alphabet[i];
      }
    }

    alphabet = uniqueAlphabet.toString();

    if (alphabet.length < MIN_ALPHABET_LENGTH) {
      throw 'alphabet must contain at least $MIN_ALPHABET_LENGTH unique characters';
    }

    if (alphabet.contains(' ')) {
      throw 'alphabet cannot contains spaces';
    }

    // seps should contain only characters present in alphabet;
    // alphabet should not contains seps
    String seps = DEFAULT_SEPS;
    for (int i = 0; i < seps.length; i++) {
      final int j = alphabet.indexOf(seps[i]);
      if (j == -1) {
        seps = seps.substring(0, i) + ' ' + seps.substring(i + 1);
      } else {
        alphabet = alphabet.substring(0, j) + ' ' + alphabet.substring(j + 1);
      }
    }

    alphabet = alphabet.replaceAll(RegExp(r'\s+|\s'), '');
    seps = seps.replaceAll(' ', '');
    seps = _consistentShuffle(seps, this.salt);

    if ((seps.isEmpty) || ((alphabet.length / seps.length) > SEP_DIV)) {
      int sepsLen = (alphabet.length / SEP_DIV).ceil();

      if (sepsLen == 1) {
        sepsLen++;
      }

      if (sepsLen > seps.length) {
        final int diff = sepsLen - seps.length;
        seps += alphabet.substring(0, diff);
        alphabet = alphabet.substring(diff);
      } else {
        seps = seps.substring(0, sepsLen);
      }
    }

    alphabet = _consistentShuffle(alphabet, this.salt);
    // use double to round up
    final int guardCount = (alphabet.length / GUARD_DIV).ceil();

    String guards;
    if (alphabet.length < 3) {
      guards = seps.substring(0, guardCount);
      seps = seps.substring(guardCount);
    } else {
      guards = alphabet.substring(0, guardCount);
      alphabet = alphabet.substring(guardCount);
    }
    this.guards = guards;
    this.alphabet = alphabet;
    this.seps = seps;
  }

  ///  Max number that can be encoded with Hashids.
  static const MAX_NUMBER = 9007199254740992;

  static const DEFAULT_ALPHABET =
      'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890';
  static const DEFAULT_SEPS = 'cfhistuCFHISTU';
  static const DEFAULT_SALT = '';

  static const int DEFAULT_MIN_HASH_LENGTH = 0;
  static const int MIN_ALPHABET_LENGTH = 16;
  static const double SEP_DIV = 3.5;
  static const int GUARD_DIV = 12;

  String salt;
  int minHashLength;
  String alphabet;
  String seps;
  String guards;

  String encode(dynamic number) {
    if (number is int) {
      return encodeInt(number);
    }
    if (number is String) {
      return encodeString(number);
    }
    if (number is List<int>) {
      return encodeList(number);
    }
    return '';
  }

  String encodeInt(int number) {
    return encodeList([number]);
  }

  String encodeString(String number) {
    final parsed = int.tryParse(number) ?? -1;
    return encodeList([parsed]);
  }

  ///
  /// Encode numbers to string
  ///
  /// @param numbers
  ///         the numbers to encode
  /// @return the encoded string
  ///
  String encodeList(List<int> numbers) {
    if (numbers?.isEmpty ?? true) {
      return '';
    }

    for (final number in numbers) {
      if (number < 0) {
        return '';
      }
      if (number > MAX_NUMBER) {
        throw 'number can not be greater than $MAX_NUMBER';
      }
    }
    return _encode(numbers);
  }

  List<int> decode(String hash) {
    if (hash.isEmpty) {
      return [];
    }

    final validChars = alphabet + guards + seps;
    for (int i = 0; i < hash.length; i++) {
      if (!validChars.contains(hash[i])) {
        return [];
      }
    }

    return _decode(hash, alphabet);
  }

  ///
  /// Encode hex to string
  ///
  /// @param hex
  ///          the hex to encode
  /// @return the encoded string
  ///
  String encodeHex(String hex) {
    final List<int> result = [];
    if (!RegExp(r'^[0-9a-fA-F]+$').hasMatch(hex)) {
      return '';
    }

    final numbers =
        RegExp(r'[\w\W]{1,12}', multiLine: true, caseSensitive: false)
            .allMatches(hex);

    for (var match in numbers) {
      result.add(int.parse('1' + match.group(0), radix: 16));
    }

    return encode(result);
  }

  ///
  /// Decode string to numbers
  ///
  /// @param hash
  ///          the encoded string
  /// @return decoded numbers
  ///
  String decodeHex(String hash) {
    var result = '';
    final List<int> numbers = decode(hash);

    for (final number in numbers) {
      result += number.toRadixString(16).substring(1);
    }

    return result.toString();
  }

  static int checkedCast(int value) {
    final int result = value;
    if (result != value) {
      // don't use checkArgument here, to avoid boxing
      throw 'Out of range: $value';
    }
    return result;
  }

  /* Private methods */

  String _encode(List<int> numbers) {
    var numberHashInt = 0;
    for (int i = 0; i < numbers.length; i++) {
      numberHashInt += numbers[i] % (i + 100);
    }
    String alphabet = this.alphabet;
    final ret = alphabet[numberHashInt % alphabet.length];

    int num;
    int sepsIndex, guardIndex;
    String buffer;
    var retStrB = '';
    retStrB += ret;
    String guard;

    for (int i = 0; i < numbers.length; i++) {
      num = numbers[i];
      buffer = ret + salt + alphabet;

      alphabet =
          _consistentShuffle(alphabet, buffer.substring(0, alphabet.length));
      final String last = _hash(num, alphabet);

      retStrB += last;

      if (i + 1 < numbers.length) {
        if (last.isNotEmpty) {
          num %= last.codeUnitAt(0) + i;
          sepsIndex = num % seps.length;
        } else {
          sepsIndex = 0;
        }
        retStrB += seps[sepsIndex];
      }
    }

    String retStr = retStrB.toString();
    if (retStr.length < minHashLength) {
      guardIndex = (numberHashInt + retStr.codeUnitAt(0)) % guards.length;
      guard = guards[guardIndex];

      retStr = guard + retStr;

      if (retStr.length < minHashLength) {
        guardIndex = (numberHashInt + retStr.codeUnitAt(2)) % guards.length;
        guard = guards[guardIndex];

        retStr += guard;
      }
    }

    final int halfLen = alphabet.length ~/ 2;
    while (retStr.length < minHashLength) {
      alphabet = _consistentShuffle(alphabet, alphabet);
      retStr =
          alphabet.substring(halfLen) + retStr + alphabet.substring(0, halfLen);
      final int excess = retStr.length - minHashLength;
      if (excess > 0) {
        final int startPos = excess ~/ 2;
        retStr = retStr.substring(startPos, startPos + minHashLength);
      }
    }

    return retStr;
  }

  List<int> _decode(String hash, String alphabet) {
    List<int> ret = <int>[];

    int i = 0;

    String hashBreakdown = hash;
    for (var j = 0; j < guards.length; j++) {
      hashBreakdown = hashBreakdown.replaceAll(guards[j], ' ');
    }
    List<String> hashArray = hashBreakdown.split(' ');

    if (hashArray.length == 3 || hashArray.length == 2) {
      i = 1;
    }

    if (hashArray.isNotEmpty) {
      hashBreakdown = hashArray[i];
      if (hashBreakdown.isNotEmpty) {
        final lottery = hashBreakdown[0];

        hashBreakdown = hashBreakdown.substring(1);
        for (var j = 0; j < seps.length; j++) {
          hashBreakdown = hashBreakdown.replaceAll(seps[j], ' ');
        }
        hashArray = hashBreakdown.split(' ');

        String subHash, buffer;
        for (final String aHashArray in hashArray) {
          subHash = aHashArray;
          buffer = lottery + salt + alphabet;
          alphabet = _consistentShuffle(
              alphabet, buffer.substring(0, alphabet.length));
          ret.add(_unhash(subHash, alphabet));
        }
      }
    }

    // transform from List<Long> to long[]

    if (encode(ret) != hash) {
      ret = [];
    }
    return ret;
  }

  static String _consistentShuffle(String alphabet, String salt) {
    int integer;

    if (salt.isEmpty) {
      return alphabet;
    }

    final tmpAlphabet = alphabet.split('');

    for (var i = tmpAlphabet.length - 1, v = 0, p = 0, j = 0; i > 0; i--, v++) {
      v %= salt.length;
      p += integer = salt.codeUnitAt(v);
      j = (integer + v + p) % i;

      final tmp = tmpAlphabet[j];
      tmpAlphabet[j] = tmpAlphabet[i];
      tmpAlphabet[i] = tmp;
    }
    return tmpAlphabet.join('');
  }

  static String _hash(int input, String alphabet) {
    String hash = '';
    final int alphabetLen = alphabet.length;

    do {
      final int index = input % alphabetLen;
      if (index >= 0 && index < alphabet.length) {
        hash = alphabet[index] + hash;
      }
      input = input ~/ alphabetLen;
    } while (input > 0);

    return hash;
  }

  int _unhash(String input, String alphabet) {
    int number = 0, pos;

    for (int i = 0; i < input.length; i++) {
      pos = alphabet.indexOf(input[i]);
      number = number * alphabet.length + pos;
    }

    return number;
  }

  ///
  /// Get Hashid algorithm version.
  ///
  /// @return Hashids algorithm version implemented.
  ///
  String getVersion() {
    return '1.0.0';
  }
}


lib/hashids2.dart
import 'dart:core';

/// A Dart class to generate YouTube-like hashes from one or many numbers.
/// {@tool sample}
///
/// The simplest use case:
///
/// ```dart
/// final hashids = HashIds();
/// final fromNumber = hashids.encode(42);
/// final number = hashids.decode(fromNumber); // [42]
/// ```
/// {@end-tool}
class HashIds {
  HashIds(
      {this.salt = DEFAULT_SALT,
      int minHashLength = DEFAULT_MIN_HASH_LENGTH,
      String alphabet = DEFAULT_ALPHABET}) {
    this.minHashLength =
        minHashLength > 0 ? minHashLength : DEFAULT_MIN_HASH_LENGTH;

    var uniqueAlphabet = '';
    for (int i = 0; i < alphabet.length; i++) {
      if (!uniqueAlphabet.contains(alphabet[i])) {
        uniqueAlphabet += alphabet[i];
      }
    }

    String alpha = uniqueAlphabet.toString();

    if (alpha.length < MIN_ALPHABET_LENGTH) {
      throw 'alphabet must contain at least $MIN_ALPHABET_LENGTH unique characters';
    }

    if (alpha.contains(' ')) {
      throw 'alphabet cannot contains spaces';
    }

    // seps should contain only characters present in alphabet;
    // alphabet should not contains seps
    String seps = DEFAULT_SEPS;
    for (int i = 0; i < seps.length; i++) {
      final int j = alpha.indexOf(seps[i]);
      if (j == -1) {
        seps = seps.substring(0, i) + ' ' + seps.substring(i + 1);
      } else {
        alpha = alpha.substring(0, j) + ' ' + alpha.substring(j + 1);
      }
    }

    alpha = alpha.replaceAll(RegExp(r'\s+|\s'), '');
    seps = seps.replaceAll(' ', '');
    seps = _consistentShuffle(seps, salt);

    if ((seps.isEmpty) || ((alpha.length / seps.length) > SEP_DIV)) {
      int sepsLen = (alpha.length / SEP_DIV).ceil();

      if (sepsLen == 1) {
        sepsLen++;
      }

      if (sepsLen > seps.length) {
        final int diff = sepsLen - seps.length;
        seps += alpha.substring(0, diff);
        alpha = alpha.substring(diff);
      } else {
        seps = seps.substring(0, sepsLen);
      }
    }

    alpha = _consistentShuffle(alpha, salt);
    // use double to round up
    final int guardCount = (alpha.length / GUARD_DIV).ceil();

    String guards;
    if (alpha.length < 3) {
      guards = seps.substring(0, guardCount);
      seps = seps.substring(guardCount);
    } else {
      guards = alpha.substring(0, guardCount);
      alpha = alpha.substring(guardCount);
    }
    this.guards = guards;
    this.alphabet = alpha;
    this.seps = seps;
  }

  ///  Max number that can be encoded with Hashids.
  static const MAX_NUMBER = 9007199254740992;

  static const DEFAULT_ALPHABET =
      'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890';
  static const DEFAULT_SEPS = 'cfhistuCFHISTU';
  static const DEFAULT_SALT = '';

  static const int DEFAULT_MIN_HASH_LENGTH = 0;
  static const int MIN_ALPHABET_LENGTH = 16;
  static const double SEP_DIV = 3.5;
  static const int GUARD_DIV = 12;

  late String salt;
  late int minHashLength;
  late String alphabet;
  late String seps;
  late String guards;

  /// Encodes int, String (with int), or list of ints
  String encode(dynamic number) {
    if (number is int) {
      return encodeInt(number);
    }
    if (number is String) {
      return encodeString(number);
    }
    if (number is List<int>) {
      return encodeList(number);
    }
    return '';
  }

  ///
  /// Encode single int
  ///
  String encodeInt(int number) {
    return encodeList([number]);
  }

  ///
  /// Encode string with number
  ///
  String encodeString(String number) {
    final parsed = int.tryParse(number) ?? -1;
    return encodeList([parsed]);
  }

  ///
  /// Encode list of numbers
  ///
  String encodeList(List<int> numbers) {
    if (numbers.isEmpty) {
      return '';
    }

    for (final number in numbers) {
      if (number < 0) {
        return '';
      }
      if (number > MAX_NUMBER) {
        throw ArgumentError('number can not be greater than $MAX_NUMBER');
      }
    }
    return _encode(numbers);
  }

  /// Decodes hasn into list of ints
  List<int> decode(String hash) {
    if (hash.isEmpty) {
      return [];
    }

    final validChars = alphabet + guards + seps;
    for (int i = 0; i < hash.length; i++) {
      if (!validChars.contains(hash[i])) {
        return [];
      }
    }

    return _decode(hash, alphabet);
  }

  ///
  /// Encode hex to string
  ///
  String encodeHex(String hex) {
    final List<int> result = [];
    if (!RegExp(r'^[0-9a-fA-F]+$').hasMatch(hex)) {
      return '';
    }

    final numbers =
        RegExp(r'[\w\W]{1,12}', multiLine: true, caseSensitive: false)
            .allMatches(hex);

    for (var match in numbers) {
      result.add(int.parse('1' + match.group(0)!, radix: 16));
    }

    return encode(result);
  }

  ///
  /// Decode string to numbers
  ///
  String decodeHex(String hash) {
    var result = '';
    final List<int> numbers = decode(hash);

    for (final number in numbers) {
      result += number.toRadixString(16).substring(1);
    }

    return result.toString();
  }

  /* Private methods */

  String _encode(List<int> numbers) {
    var numberHashInt = 0;
    for (int i = 0; i < numbers.length; i++) {
      numberHashInt += numbers[i] % (i + 100);
    }
    String alphabet = this.alphabet;
    final ret = alphabet[numberHashInt % alphabet.length];

    int num;
    int sepsIndex, guardIndex;
    String buffer;
    var retStrB = '';
    retStrB += ret;
    String guard;

    for (int i = 0; i < numbers.length; i++) {
      num = numbers[i];
      buffer = ret + salt + alphabet;

      alphabet =
          _consistentShuffle(alphabet, buffer.substring(0, alphabet.length));
      final String last = _hash(num, alphabet);

      retStrB += last;

      if (i + 1 < numbers.length) {
        if (last.isNotEmpty) {
          num %= last.codeUnitAt(0) + i;
          sepsIndex = num % seps.length;
        } else {
          sepsIndex = 0;
        }
        retStrB += seps[sepsIndex];
      }
    }

    String retStr = retStrB.toString();
    if (retStr.length < minHashLength) {
      guardIndex = (numberHashInt + retStr.codeUnitAt(0)) % guards.length;
      guard = guards[guardIndex];

      retStr = guard + retStr;

      if (retStr.length < minHashLength) {
        guardIndex = (numberHashInt + retStr.codeUnitAt(2)) % guards.length;
        guard = guards[guardIndex];

        retStr += guard;
      }
    }

    final int halfLen = alphabet.length ~/ 2;
    while (retStr.length < minHashLength) {
      alphabet = _consistentShuffle(alphabet, alphabet);
      retStr =
          alphabet.substring(halfLen) + retStr + alphabet.substring(0, halfLen);
      final int excess = retStr.length - minHashLength;
      if (excess > 0) {
        final int startPos = excess ~/ 2;
        retStr = retStr.substring(startPos, startPos + minHashLength);
      }
    }

    return retStr;
  }

  List<int> _decode(String hash, String alphabet) {
    List<int> ret = <int>[];

    int i = 0;

    String hashBreakdown = hash;
    for (var j = 0; j < guards.length; j++) {
      hashBreakdown = hashBreakdown.replaceAll(guards[j], ' ');
    }
    List<String> hashArray = hashBreakdown.split(' ');

    if (hashArray.length == 3 || hashArray.length == 2) {
      i = 1;
    }

    if (hashArray.isNotEmpty) {
      hashBreakdown = hashArray[i];
      if (hashBreakdown.isNotEmpty) {
        final lottery = hashBreakdown[0];

        hashBreakdown = hashBreakdown.substring(1);
        for (var j = 0; j < seps.length; j++) {
          hashBreakdown = hashBreakdown.replaceAll(seps[j], ' ');
        }
        hashArray = hashBreakdown.split(' ');

        String subHash, buffer;
        for (final String aHashArray in hashArray) {
          subHash = aHashArray;
          buffer = lottery + salt + alphabet;
          alphabet = _consistentShuffle(
              alphabet, buffer.substring(0, alphabet.length));
          ret.add(_unhash(subHash, alphabet));
        }
      }
    }

    // transform from List<Long> to long[]

    if (encode(ret) != hash) {
      ret = [];
    }
    return ret;
  }

  static String _consistentShuffle(String alphabet, String salt) {
    int integer;

    if (salt.isEmpty) {
      return alphabet;
    }

    final tmpAlphabet = alphabet.split('');

    for (var i = tmpAlphabet.length - 1, v = 0, p = 0, j = 0; i > 0; i--, v++) {
      v %= salt.length;
      p += integer = salt.codeUnitAt(v);
      j = (integer + v + p) % i;

      final tmp = tmpAlphabet[j];
      tmpAlphabet[j] = tmpAlphabet[i];
      tmpAlphabet[i] = tmp;
    }
    return tmpAlphabet.join('');
  }

  static String _hash(int input, String alphabet) {
    String hash = '';
    final int alphabetLen = alphabet.length;

    do {
      final int index = input % alphabetLen;
      if (index >= 0 && index < alphabet.length) {
        hash = alphabet[index] + hash;
      }
      input = input ~/ alphabetLen;
    } while (input > 0);

    return hash;
  }

  int _unhash(String input, String alphabet) {
    int number = 0, pos;

    for (int i = 0; i < input.length; i++) {
      pos = alphabet.indexOf(input[i]);
      number = number * alphabet.length + pos;
    }

    return number;
  }
}
